File: server/inscode/encode.py
"""Text -> QR matrix, and the bit-packing used to ship a matrix inside a QRSpec."""
import base64

import qrcode
from qrcode.constants import ERROR_CORRECT_H, ERROR_CORRECT_L, ERROR_CORRECT_M, ERROR_CORRECT_Q

from .spec import Encoded, QRSpec

_EC = {"L": ERROR_CORRECT_L, "M": ERROR_CORRECT_M, "Q": ERROR_CORRECT_Q, "H": ERROR_CORRECT_H}


class Matrix:
    """A decoded matrix. Reserved cells come from geometry.py, not from here, so
    that both renderers derive them from the same rules."""

    __slots__ = ("size", "version", "mask_pattern", "_bits")

    def __init__(self, size: int, version: int, mask_pattern: int, bits: bytes):
        self.size = size
        self.version = version
        self.mask_pattern = mask_pattern
        self._bits = bits

    def get(self, row: int, col: int) -> bool:
        if not (0 <= row < self.size and 0 <= col < self.size):
            return False
        i = row * self.size + col
        return bool(self._bits[i >> 3] & (0x80 >> (i & 7)))


def _pack(size: int, rows) -> bytes:
    out = bytearray((size * size + 7) // 8)
    for r in range(size):
        for c in range(size):
            if rows[r][c]:
                i = r * size + c
                out[i >> 3] |= 0x80 >> (i & 7)
    return bytes(out)


def encode(text: str, ec_level: str = "H", mask_pattern: int | None = None) -> Encoded:
    """Encode text. Pass `mask_pattern` to reproduce an earlier encode exactly."""
    qr = qrcode.QRCode(
        error_correction=_EC[ec_level],
        border=0,  # the quiet zone is added at render time, as in the TS side
        mask_pattern=mask_pattern,
    )
    qr.add_data(text or " ")
    qr.make(fit=True)
    if mask_pattern is None:
        # python-qrcode picks a mask but never records which. best_mask_pattern()
        # recomputes it -- and leaves the last *test* mask in qr.modules as a side
        # effect, so the chosen one has to be re-applied before reading the matrix.
        mask_pattern = qr.best_mask_pattern()
        qr.makeImpl(False, mask_pattern)
    rows = qr.get_matrix()
    size = len(rows)
    chosen = mask_pattern
    return Encoded(
        version=qr.version,
        maskPattern=int(chosen),
        size=size,
        bits=base64.b64encode(_pack(size, rows)).decode("ascii"),
    )


def matrix_from_encoded(enc: Encoded) -> Matrix:
    bits = base64.b64decode(enc.bits)
    need = (enc.size * enc.size + 7) // 8
    if len(bits) < need:
        raise ValueError(f"encoded.bits too short: {len(bits)} < {need}")
    return Matrix(enc.size, enc.version, enc.mask_pattern, bits)


def matrix_for(spec: QRSpec) -> Matrix:
    """The matrix for a spec. `spec.encoded` wins when present."""
    enc = spec.encoded or encode(spec.content.text, spec.content.ec_level)
    return matrix_from_encoded(enc)

File: web/src/qr/encode.ts
/**
 * Text -> QR matrix, plus the bit-packing used to ship a matrix inside a QRSpec.
 */
import QRCode from 'qrcode'
import type { ECLevel, Encoded, QRSpec } from './spec'

/** Matrix width in modules for a version, quiet zone excluded. */
export const symbolSize = (version: number) => version * 4 + 17

/**
 * A decoded matrix. `get` is the only accessor the renderer needs -- reserved
 * cells come from geometry.ts so that the browser and the Python API derive them
 * the same way, from the same rules.
 */
export interface Matrix {
  size: number
  version: number
  maskPattern: number
  get(row: number, col: number): boolean
}

function packBits(size: number, get: (r: number, c: number) => boolean): string {
  const bytes = new Uint8Array(Math.ceil((size * size) / 8))
  for (let r = 0; r < size; r++) {
    for (let c = 0; c < size; c++) {
      if (!get(r, c)) continue
      const i = r * size + c
      bytes[i >> 3] |= 0x80 >> (i & 7)
    }
  }
  let s = ''
  for (const b of bytes) s += String.fromCharCode(b)
  return btoa(s)
}

function unpackBits(size: number, b64: string): Uint8Array {
  const bin = atob(b64)
  const bytes = new Uint8Array(bin.length)
  for (let i = 0; i < bin.length; i++) bytes[i] = bin.charCodeAt(i)
  const need = Math.ceil((size * size) / 8)
  if (bytes.length < need) throw new Error(`encoded.bits too short: ${bytes.length} < ${need}`)
  return bytes
}

/** node-qrcode types this 0..7 union; we validate the number at the boundary instead. */
type MaskPattern = NonNullable<Parameters<typeof QRCode.create>[1]>['maskPattern']

/** Encode text into a matrix. Pass `maskPattern` to reproduce an earlier encode exactly. */
export function encode(text: string, ecLevel: ECLevel, maskPattern?: number): Encoded {
  const qr = QRCode.create(text || ' ', {
    errorCorrectionLevel: ecLevel,
    ...(maskPattern === undefined ? {} : { maskPattern: maskPattern as MaskPattern }),
  })
  // Every renderer must reproduce this exact mask, so an unknown one is fatal
  // rather than something to paper over with a default.
  if (typeof qr.maskPattern !== 'number') {
    throw new Error('qrcode did not report a mask pattern; cannot guarantee render parity')
  }
  const size = qr.modules.size
  return {
    version: qr.version,
    maskPattern: qr.maskPattern,
    size,
    bits: packBits(size, (r, c) => Boolean(qr.modules.get(r, c))),
  }
}

export function matrixFromEncoded(enc: Encoded): Matrix {
  const { size, version, maskPattern } = enc
  const bytes = unpackBits(size, enc.bits)
  return {
    size,
    version,
    maskPattern,
    get(r, c) {
      if (r < 0 || c < 0 || r >= size || c >= size) return false
      const i = r * size + c
      return (bytes[i >> 3] & (0x80 >> (i & 7))) !== 0
    },
  }
}

/**
 * The matrix for a spec. `spec.encoded` wins when present -- it is the browser's
 * own encode result travelling with the design so every renderer draws the same
 * bits. Only re-encode when it is absent (direct API use).
 */
export function matrixFor(spec: QRSpec): Matrix {
  const enc = spec.encoded ?? encode(spec.content.text, spec.content.ecLevel)
  return matrixFromEncoded(enc)
}
